src/lib/png/png.c

```diff
@@ -0,0 +1,976 @@
+	tree->tree2d = 0;
+	tree->tree1d = 0;
+	tree->lengths = 0;
+}
+
+static void HuffmanTree_cleanup(HuffmanTree* tree) {
+	free(tree->tree2d);
+	free(tree->tree1d);
+	free(tree->lengths);
+}
+
+/* fun to decode symbols, return decoded symbol or error(-1) */
+static int huffman_decode_symbol(uint8_t* in, int* bp,
+				HuffmanTree* tree, int inbitlength) {
+	int pos = 0, tmp;
+	/* endless 'for' until get symbol or get error */
+	for (;;) {
+		if (*bp >= inbitlength) {
+			return -1; /* error: end of input memory reached without endcode */
+		}
+		tmp = tree->tree2d[(pos << 1) + READBIT(*bp, in)];
+		(*bp)++;
+
+		if (tmp < tree->num_of_codes) {
+			return tmp; /*get decoded symbol, return it*/
+		} else {
+			pos = tmp - tree->num_of_codes; /* not yet symbol, continue */
+		}
+
+		if (pos >= tree->num_of_codes) {
+			return -1; /* error: it appeared outside the codetree */
+		}
+	}
+}
+
+/*
+ * make from tree1d[] tree2d[][]. value = 6666 means not init
+ * value < number_of_codes is a codes
+ * value >= number_of_codes is an address to another bit
+ * The 2 rows are the 2 possible bit values (0 or 1), there are as
+ * many columns as codes - 1.
+ */
+static int make_tree2d(HuffmanTree *tree) {
+	int nodefill_info = 0;
+	int pos = 0; /* pos in the tree */
+	int i, n;
+
+	tree->tree2d = (int*)malloc(tree->num_of_codes * 2 * sizeof(int));
+	if (!tree->tree2d) {
+		return ENOMEM; /* malloc fail */
+	}
+
+	for (n = 0; n < tree->num_of_codes * 2; n++) {
+		tree->tree2d[n] = 6666; /* 6666 means not filled */
+	}
+
+	for (n = 0; n < tree->num_of_codes; n++) /* the codes */ {
+		for (i = 0; i != tree->lengths[n]; i++) /* the bits for this code */ {
+
+			uint8_t bit = (uint8_t)((tree->tree1d[n] >> (tree->lengths[n] - i - 1)) & 1);
+			if (tree->tree2d[2 * pos + bit] == 6666) /* not yet filled */ {
+				if (i + 1 == tree->lengths[n]) /* last bit */ {
+					tree->tree2d[2 * pos + bit] = n; /* put the current code in it */
+					pos = 0;
+				} else {
+					/* put address of the next step, it's (nodefilled + 1) */
+					nodefill_info++;
+					/* addresses encoded with numcodes added to it */
+					tree->tree2d[2 * pos + bit] = nodefill_info + tree->num_of_codes;
+					pos = nodefill_info;
+				}
+			} else {
+				pos = tree->tree2d[2 * pos + bit] - tree->num_of_codes;
+			}
+		}
+	}
+	for (n = 0; n < tree->num_of_codes * 2; n++) {
+		if (tree->tree2d[n] == 6666) {
+			tree->tree2d[n] = 0; /* remove possible remaining 6666 */
+		}
+	}
+	return 0;
+}
+
+/*
+ * Second step of making tree.
+ * num_of_codes, lengths need to be filled.
+ * return 0 with success or error.
+ */
+static int make_tree_second_step(HuffmanTree *tree) {
+	int blcount[tree->maxbitlen + 1];
+	int nextcode[tree->maxbitlen + 1];
+	int bits, i, err = 0;
+
+	for (i = 0; i < tree->maxbitlen + 1; i++) {
+		blcount[i] = 0;
+		nextcode[i] = 0;
+	}
+
+	tree->tree1d = (int*)malloc(tree->num_of_codes * sizeof(int));
+	if (!tree->tree1d) {
+		err = ENOMEM; /* malloc fail */
```

Comment:
Is this error handled somehow? May be add `log_error()`?

Reply:
done

Reply:
add log_error in all lib

Reply:
I've reworked the file reading algorithm and now it can decode and draw all the pictures I've tested
also added depend in fb_draw_image and added fb_draw_image in platform/nuklear/arm_qemu template